proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+Retriever -- "documents (List[List[Doc]])"  --> Prompts
+Prompts -- "prompts (List[str])" --> GPT4
+GPT4 -- "replies (List[List[str]])" --> RepliesToAnswersConverter
+RepliesToAnswersConverter -- "answers (List[List[Answer]])" --> OUT{OUT}
+```
+
+While the code for such pipeline may look like:
+
+```python
+from haystack.preview.components import MemoryRetriever, Prompts, ChatGPT, RepliesToAnswersConverter
+from haystack.preview.document_stores import MemoryDocumentStore
+from haystack.preview.pipeline import Pipeline
+
+pipe = Pipeline()
+pipe.add_store("store", MemoryDocumentStore())
+pipe.add_component("retriever", MemoryRetriever(), store="store")
+pipe.add_component("prompts", Prompts("deepset/question-answering"))
+pipe.add_component("llm", GPT4(api_key="..."))
+pipe.add_component("replies_converter", RepliesToAnswersConverter())
+
+pipe.connect("retriever", "prompt")
+pipe.connect("prompt", "llm")
+pipe.connect("llm", "replies_converter")
+
+questions = ["Why?", "Why not?"]
+results = pipe.run({
+	"retriever": {"queries": questions},
+	"prompt": {"questions": questions},
+})
+
+assert results == {
+	"replies_converter": {
+    "answers": [[Answer("Because of this.")], [Answer("Because of that.")]]
+  }
+}
+```
+
+# Detailed design
+
+Haystack’s `PromptNode` is a very complex component that includes under its name several functionalities: loading prompt templates through the `PromptTemplate` class, rendering such prompt template with the variables from the invocation context, choosing which LLM backend to use, sending the prompt to the LLM using the correct invocation layer, interpreting the results, parsing them into objects, and putting them back in the pipeline in a way other components can understand.
+
+in Haystack 2.0 we unpack these functionalities into a few separate components, to clarify what is happening, how it works, and provide additional flexibility.
+
+The main functionalities we identified are the following:
+
+1. Fetching the prompt from different sources
+2. Rendering the prompt using variables
+3. Invoke the LLM
+4. Parse the output
+5. History/Memory management
+
+We leave the discussion about History/Memory to a separate proposal, as it concerns mostly Agents, and focus on the other points.
+
+## LLM invocation
+
+In Haystack 1.x, `PromptNode` uses `InvocationLayer` to query different LLMs under a unified API. In that design, users do not need to know which invocation layer is used for the model they select, as `PromptNode` takes responsibility of selecting it.
+
+Such invocation layers can be ported to 2.0 as standalone components. In this way we will have one component for each LLM backed that we support.
+
+Each component should be named after the class of models it supports. For example we should have `GPT4`, `ChatGPT`, etc. For components whose name may be confusing, or that include models that are not LLMs, we can add a LLM suffix: like `FalconLLM`, `CohereLLM`, `HuggingFaceLLM`.
```

Comment:
Will chat and foundation models be split? What about different context lengths, will `GPT-4` support gpt-4-32k and gpt-4-16k and so on? What about model snapshots?

Reply:
Don't take this list at face value: we'll take the precise decision of which exact components to have at implementation time. It will mostly depend on how similar their implementation are and how many parameters they end up having.

Reply:
Moved to Notion!